concepts/for-loops/introduction.md

```diff
@@ -1,54 +1,69 @@
+The `initialization` sets an initial state for the loop and is executed exactly once at the start of the loop.
+Typically it declares and assigns a variable used in the test expression and update statement.
+For example:
 
 ```java
-// Declare array with explicit size (size is 2)
-int[] twoInts = new int[2];
+int i = 1
+```
 
-// Assign second element by index
-twoInts[1] = 8;
+The `test` expression tests if the loop should end.
+If it evaluates to true, the body and then the update expression will be executed.
+If it evaluates to false, neither the body nor the update statement will be executed and execution resumes after the loop's closing bracket.
+Typically it checks the variable assigned in the initialization block.
+For example:
 
-// Retrieve the second element by index and assign to the int element
-int element = twoInts[1];
+```java
+i <= 10
 ```
 
-Arrays can also be defined using a shortcut notation that allows you to both create the array and set its value. As the compiler can now tell how many elements the array will have, the length can be omitted:
+After executing the loop body, the `update` expression is executed.
+Typically it increments or decrements the loop variable by some value.
+For example:
 
 ```java
-// Two equivalent ways to declare and initialize an array (size is 3)
-int[] threeIntsV1 = new int[] { 4, 9, 7 };
-int[] threeIntsV2 = { 4, 9, 7 };
+i++
 ```
 
-Arrays can be manipulated by either calling an array instance's methods or properties, or by using the static methods defined in the `Array` class.
-
-The fact that an array is also a _collection_ means that, besides accessing values by index, you can iterate over _all_ its values using a `foreach` loop:
+A for loop printing out the first four squares would look like this:
 
 ```java
-char[] vowels = { 'a', 'e', 'i', 'o', 'u' };
-
-for(char vowel:vowels) {
-    // Output the vowel
-    System.out.print(vowel);
+for (int i = 1; i <= 4; i++) {
+    System.out.println("square of " + i + " is " + i * i);
 }
+```
+
+The output would be:
 
-// => aeiou
+```
+square of 1 is 1
+square of 2 is 4
+square of 3 is 9
+square of 4 is 16
 ```
 
-If you want more control over which values to iterate over, a `for` loop can be used:
+If iterating through every element in a collection, a for-each loop is preferred, but it can be done with a for loop like this:
 
 ```java
-char[] vowels = { 'a', 'e', 'i', 'o', 'u' };
-
-for (int i = 0; i < 3; i++) {
-    // Output the vowel
-    System.out.print(vowels[i]);
+for (int i = 0; i < array.length; i++) {
+    System.out.print(array[i]);
 }
-
-// => aei
 ```
+
+A `for` loop does have some advantages over a `for-each` loop:
+
+- You can start or stop at the index you want.
+- You can use any (boolean) termination condition you want.
```

Comment:
```suggestion
- You can use any (Boolean) termination condition you want.
```

Reply:
Not sure I agree with this one. Using the uppercase Boolean makes it seem like you are referring specifically to the class, but any primitive boolean would do just as well.

Reply:
It is a persons name, and in English, at least, Boolean is what I believe is used.  In Ruby a generic "boolean" would be anything, with only `nil` and `false` not evaluating as truthy.

In Ruby, it is the only place I do not use a capital "B" for Boolean when I am talking about Boolean in the less strict sense as we can because of the truthy/falsey nature that is allowed, otherwise I have used Boolean capitalized since gradeschool, and though electronics schooling.

I may be the exception though.  But I do not think so.

[This English SE entry](https://english.stackexchange.com/a/4491/83340) has more authoritative references than I would contain as an English as a Not First Language person.

Reply:
Actually that is technically incorrect. The name of the individual for which booleans are named is George Boole.

In Java specifically, Boolean refers to the boxed class for the primitive boolean (like Integer refers to the boxed class for the primitive int). Therefore, I think boolean should be used because we do not specifically mean the boxed Boolean class.

Reply:
I can code-font `boolean` if that would help clarify things.